src/server/main_service.cc

```diff
@@ -1047,6 +1047,34 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx)
   }
 }
 
+class ReplyGuard {
+ public:
+  ReplyGuard(ConnectionContext* cntx, std::string_view cid_name) : cntx(cntx) {
+    const bool is_script = bool(cntx->conn_state.script_info);
+    const bool is_redis = dynamic_cast<RedisReplyBuilder*>(cntx->reply_builder()) != nullptr;
+    const bool is_one_of =
+        absl::flat_hash_set<std::string_view>({"REPLCONF", "DFLY", "EXEC"}).contains(cid_name);
+    const bool should_dcheck = !is_one_of && !is_script && is_redis;
+    if (should_dcheck) {
+      old = static_cast<RedisReplyBuilder*>(cntx->Inject(&crb));
+    }
+  }
+
+  ~ReplyGuard() {
+    if (old) {
+      auto reply = crb.Take();
+      DCHECK(reply.index() > 0);
+      CapturingReplyBuilder::Apply(std::move(reply), old);
```

Comment:
much nicer! My question is - could it be that injecting  CapturingReplyBuilder would actually change the debug version significantly from the release in terms of I/O pattern?

Reply:
somewhat, now replies are sent only once the command finished executing

Reply:
It does delay the replies after the invocation of the command. So yes, it does alter the behavior. Originally, I had it implemented differently (see the early commits) but the CapturingReplyBuilder seems a clearer approach.

> would actually change the debug version significantly from the release in terms of I/O pattern?
I think it also changes the functionality, as I described above which I agree it is not great but in this context maybe it's ok if you take into account that the release version is also different because of different optimizations. 

There is another solution in the previous commits (which can be cleaned up as well) but this also pollutes the `RedisReplyBuilder` (but it doesn't delay the replies). What's your take?

Reply:
I have not see your solution but is not it just about setting a boolean in a few low-level functions?

Reply:
i mean in replybuilder ? I am ok with this solution but I do preference with poluting replybuilder if it is simple and efficient.